File: elixir/meetup/meetup.exs
defmodule Meetup do
  @moduledoc """
  Calculate meetup dates.
  """

  @type weekday ::
      :monday | :tuesday | :wednesday
    | :thursday | :friday | :saturday | :sunday

  @type schedule :: :first | :second | :third | :fourth | :last | :teenth

  @doc """
  Calculate a meetup date.

  The schedule is in which week (1..4, last or "teenth") the meetup date should
  fall.
  """
  @spec meetup(pos_integer, pos_integer, weekday, schedule) :: :calendar.date
  def meetup(year, month, weekday, schedule) do
    days_in_month(year, month)
    |> Enum.filter(&(to_weekday(Date.day_of_week(&1)) == weekday))
    |> day_for_schedule(schedule)
    |> Date.to_erl
  end

  def days_in_month(year, month) do
    num_days = Date.days_in_month(Date.from_erl!({year, month, 1}))
    Enum.map(1..num_days, &(Date.from_erl!({year, month, &1})))
  end

  def day_for_schedule(weekdays, :first), do: Enum.at(weekdays, 0)
  def day_for_schedule(weekdays, :second), do: Enum.at(weekdays, 1)
  def day_for_schedule(weekdays, :third), do: Enum.at(weekdays, 2)
  def day_for_schedule(weekdays, :fourth), do: Enum.at(weekdays, 3)
  def day_for_schedule(weekdays, :last), do: Enum.at(weekdays, -1)
  def day_for_schedule(weekdays, :teenth), do: Enum.find(weekdays, &(&1.day in 13..19))

  @weekdays %{
    1 => :monday, 2 => :tuesday,  3 => :wednesday, 4 => :thursday,
    5 => :friday, 6 => :saturday, 7 => :sunday
  }

  def to_weekday(day_num), do: @weekdays[day_num]
end

File: python/meetup/meetup.py
from calendar import Calendar
from functools import partial

weekday_to_ord = {'Monday': 0, 'Tuesday': 1, 'Wednesday': 2, 'Thursday': 3, 'Friday': 4, 'Saturday': 5, 'Sunday': 6}

def matching_weekdays(year, month, weekday):
  return [date
          for date in Calendar().itermonthdates(year, month)
          if date.month == month and date.weekday() == weekday_to_ord[weekday]]

def meetup_day(year, month, weekday, schedule):
  def nth(n, dates):
    return dates[n]

  def last(dates):
    return dates[-1]

  def teenth(dates):
    return next(date for date in dates if date.day in range(13, 20))

  select_day = {
      '1st': partial(nth, 0),
      '2nd': partial(nth, 1),
      '3rd': partial(nth, 2),
      '4th': partial(nth, 3),
      '5th': partial(nth, 4),
      'last': last,
      'teenth': teenth
  }

  return select_day[schedule](matching_weekdays(year, month, weekday))
